### ISSUE TITLE
ConfigurationItem on RFC throws SystemNullRefrence Exception

### ISSUE THREAD

**Issue by RobertOsterberg** [code4me/4me-sdk-csharp#53]
I am trying to get a ConfigurationItem from a RFC in Amelia, but the api does not find it thus the exception.

the request object contains ConfigurationItem but this property is also NULL
string computerName = request.ConfigurationItem.Name; (exception is thrown)

Any suggestion on how to retrive the CI? i have tried to read the documentation but it does not include how to get linked items from a request.

Have a great day

**Comment by klaasvandeweerdt** [comment]
@RobertOsterberg, The CI is not included in the [default response](https://developer.4me.com/v1/requests/#collection-fields).
The ```Get``` method allows you to specify the field names. The field name for the ```ConfigurationItem``` property is ```ci```.

This is an example
```List<Request> requests = client.Requests.Get(new Filter("id", FilterCondition.Equality, 111), "name", "ci");```
Once you specify a field name the defaults are ignore, so make sure to add all field names required.

A request can have more than one configuration item linked to it. The ```request.ConfigurationItem``` will only return this first one.
You can get all linked configuration items via a second request referencing the result of the first response.
```List<ConfigurationItem> cis = client.Requests.GetConfigurationItems(requests[0]);```

Hope this helps

**Comment by RobertOsterberg** [comment]
@klaasvandeweerdt - Thanks for you quik response, and Yes i got the CIS full list of properties using 
List<ConfigurationItem> cis = client.Requests.GetConfigurationItems(requests[0]); this line. 

That should be in the documentation

Have a great day

**Comment by RobertOsterberg** [comment]
Hi again
i am trying to get to the details of a request 
![image](https://user-images.githubusercontent.com/18688250/172208691-50b2011d-d2fe-480d-9c1b-3960523931de.png)

Also i would like to know how to get a specifik activity of an automation WF within a request if that is possible. 

Good work keep it up

**Comment by klaasvandeweerdt** [comment]
@RobertOsterberg,

First for the configuration items. I see if I can add it to the readme file. This logic applies to almost all "one to many" relations, this data needs to be retrieved via a second API call. All those relations can be found on the [4me REST API](https://developer.4me.com/v1/) developer pages.

I assume that the details screenshot is a request with UI extension and custom fields. Those values are stored in the field ```custom_fields```. The custom fields property is exposed on all data entities supporting custom fields as a ```CustomFieldCollection```, which is an implementation of ```IDictionary<string, string>```. You can access them via ```request[0].CustomField["custom_field_name"]```.

The [Automation Rules End point](https://developer.4me.com/v1/automation_rules/) is not implemented in the SDK. For what I can see it allows you to manage automation rules not retrieve a status. I guess there would be other ways to get to the same result. Specific values in the requests, such as custom fields; or the status, a related workflow value in the request ```workflow```.
Another way could be storing a specific value in the custom_field by the automation rule, upon creation give it the value 0 once the automation was executed set it to 1.

**Comment by RobertOsterberg** [comment]
Hi i am not sure how to get to the CustomField. I have tried this IDictionary<string, string> customFields = request.CustomFields; but no result the way you specified above did not work either.

If you wanted to get the Value of the Key in the CollectionField above screenshot say Key="What type of computer" and then get the value how do you do that?

i tried this IDictionary<string, string> customFields = request.CustomFields["What type of Computer"]; with no luck

Thx

**Comment by klaasvandeweerdt** [comment]
Hi, 
When a requested is returned via a filter it will return only the [default fields](https://developer.4me.com/v1/requests/#collection-fields). In that case you need to specify which fields to return for the request. This example allows you to loop through all known custom field values.
```csharp
List<Request> requests = client.Requests.Get(new Filter("id", FilterCondition.Equality, 123), "subject", "ci", "custom_fields");
CustomFieldCollection customFields = requests[0].CustomFields;
foreach (string key in customFields.Keys)
    Console.WriteLine($"{key} : {customFields[key]}");
```
It will get all requests where the ```ID``` is 123 and load the ```subject```, ```ci``` and ```custom fields``` values. It stores the custom fields in a custom dictionary object, ```CustomFieldCollection```.
4me formats the key by default in a snake case format, so the custom field key for "What type of computer" will probably be "what_type_of_computer". You can find the exact name via the key values of the CustomField property or via the fields when editing an UI extension.

This is an example of a default text box in a UI extension.
```html
<div class="row vertical">
  <label for="what_type_of_computer" title="What type of computer">What type of computer</label>
  <input id="what_type_of_computer" type="text" autocomplete="off" class="required">
</div>
```

**Comment by RobertOsterberg** [comment]
Thank you !! that worked great! i used this to get the whole collection of keys and values.

`List<Request> details = client.Requests.Get(new Filter("ID", FilterCondition.Equality, request.ID), "custom_fields");
                                CustomFieldCollection cf = details[0].CustomFields;
                                foreach (string key in cf.Keys)
                                {
                                    Console.WriteLine($"Key:{key} - customField: {cf[key]} ");
                                }`

hope i don? bother you to much, keep up the good work.

**Comment by RobertOsterberg** [comment]
Hi again
when trying to update the request i got the data from i get this error
`Sdk4me.Sdk4meException: 'The remote server returned an error: (422) Unprocessable Entity.'`

code i am using
`Request updReq = client.Requests.Get(request.ID);
                                updReq.Status = RequestStatus.Completed;
                                updReq = client.Requests.Update(updReq);`
is this because of permission on the user account or anything else that i missed

thanks again

**Comment by klaasvandeweerdt** [comment]
Hi again as well,
[Error 422](https://developer.4me.com/v1/#422-unprocessable-entity---semantic-error-in-provided-data) means semantic error in provided data, so the request cannot be processed. I'm surprised by the error as the SDK is supposed to read the response message and return the actual error. I will have a look at that.

Question, which version of dot net are you using?

Looking at the code you are missing at least the note. Setting a request to completed requires a note.
```
Request updReq = client.Requests.Get(request.ID);
updReq.Status = RequestStatus.Completed;
updReq.Note = "A reason to complete the request";
updReq = client.Requests.Update(updReq);
```

**Comment by RobertOsterberg** [comment]
I am using net core 3.1 on this test project.
And you are correct you need to put in a note to be able to Complete the request.

I will try it tomorrow and let you you know

Thx again for you support!

have a nice evening

**Comment by klaasvandeweerdt** [comment]
Thanks for the info. I will have a look why the it return the 422 and not the message.
You will have to set all field value like you would do in the UI (same fields are required for completing a request). So completion reason will also be required.
Have a good evening as well

**Comment by RobertOsterberg** [comment]
Hi
i found out that there is one more mandatory field when completing the request and it is Completion reason so the code that i used is this
` updReq = client.Requests.Get(request.ID);
                                updReq.Note = "A reason to complete the request";
                                updReq.CompletionReason = RequestCompletionReason.Solved;
                                updReq.Status = RequestStatus.Completed;
                                updReq = client.Requests.Update(updReq);`

the note updated fine even when setting status = In Progress

thx again

**Comment by klaasvandeweerdt** [comment]
@RobertOsterberg,
I published a new [NuGet v2.0.3](https://www.nuget.org/packages/Sdk4me/).
It includes the bugfix for the missing details in the exception message and XML documentation file.

**Comment by RobertOsterberg** [comment]
> @RobertOsterberg, I published a new [NuGet v2.0.3](https://www.nuget.org/packages/Sdk4me/). It includes the bugfix for the missing details in the exception message and XML documentation file.

I will update and check it out!
